Translate the labels for units of measure

coreymckrill opened the pull request:
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

### Changes proposed in this Pull Request:

Introduces data and methods for localizing the labels on units of measure, e.g. `kg` or `cm`, and ensures they are utilized wherever those labels are output on the front end and in the admin UI.

Closes #31486

- [ ] This PR is a very minor change/addition and does not require testing instructions (if checked you can ignore/remove the next section).

### How to test the changes in this Pull Request:

Testing this involves looking through the UI (front end and admin) to see if there is anywhere that still shows untranslated unit measure labels. In order to differentiate from labels that _have_ been translated, you can add this snippet as an mu-plugin on your test site:

```php
<?php

add_filter(
	'gettext',
	function( $translated, $original ) {
		$weight_units    = array( 'kg', 'g', 'lbs', 'oz' );
		$dimension_units = array( 'm', 'cm', 'mm', 'in', 'yd' );

		$new_string = $original;

		if ( in_array( $new_string, $weight_units, true ) ) {
			$new_string = "💰";
		} else if ( in_array( $new_string, $dimension_units, true ) ) {
			$new_string = "🦄";
		}

		return $new_string;
	},
	99,
	2
);
```

This will make it so any abbreviated weight unit label that's successfully being translated will show as 💰 and any translated dimensions label will show as 🦄.

Some places to check:
* Single product views
* Product management screens
* Product import/export
* Shipping management screens
* WooCommerce settings

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> changelog add`?

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

coreymckrill committed:
Add i18n file for units

coreymckrill committed:
Add I18nUtil class

coreymckrill committed:
Update weight/dim formatting to use translated labels

coreymckrill committed:
Add changelog file

coreymckrill committed:
Update products settings to use I18nUtil for unit labels

coreymckrill committed:
Update labels for weight units

coreymckrill committed:
Update remaining labels for dimensions units

coreymckrill committed:
Add unit tests

coreymckrill committed:
Fix linting issues

barryhughes commented on plugins/woocommerce/includes/rest-api/Controllers/Version3/class-wc-rest-products-controller.php:
```diff
@@ -804,9 +806,9 @@ protected function prepare_object_for_database( $request, $creating = false ) {
 	 */
 	public function get_item_schema() {
-		$weight_unit    = get_option( 'woocommerce_weight_unit' );
-		$dimension_unit = get_option( 'woocommerce_dimension_unit' );
-		$schema         = array(
+		$weight_unit_label    = I18nUtil::get_weight_unit_label( get_option( 'woocommerce_weight_unit', 'kg' ) );
+		$dimension_unit_label = I18nUtil::get_dimensions_unit_label( get_option( 'woocommerce_dimension_unit', 'cm' ) );
+		$schema               = array(
```
Back-compat question (not necessarily a blocker, but wanted to pose the question): supposing a merchant is unhappy with the default weight units as supported by WooCommerce (`kg` | `g` | `lbs` | `oz`) and prefers to use something appropriate to their use case, like the [`dwt` pennyweight](https://en.wikipedia.org/wiki/Pennyweight) and, to do this, they update the `woocommerce_weight_unit` option directly.

Is there a risk of creating a minor breakage in that, previously, this would work:

```php
get_option( 'woocommerce_weight_unit' ); # => 'dwt'
```

Whereas now, I think they would get back an empty string (at least, if I'm reading [this code](https://github.com/woocommerce/woocommerce/pull/36708/files#diff-e1d2d755cf727d32dc1c0e14d667f66e96201ce68cf78a2f6b3973e7b93ee1cbR31-R37) correctly)?

coreymckrill replied:
Good point. Perhaps it would be better for the label to default to the unit value ("dwt" in this case) instead of an empty string?

barryhughes replied:
I think that might be the best course of action.

(It crossed my mind that we could additionally make the result filterable: but perhaps it is better to hold on that until a concrete request for such a thing surfaces—rather than prematurely increase the public API surface.)

coreymckrill replied:
Updated in b336652e0d46bfa96c1257e5bd2424e25e07d632

barryhughes commented on plugins/woocommerce/i18n/units.php:
```diff
@@ -0,0 +1,28 @@
+
+return array(
+	'weight'     => array(
+		'kg'  => _x( 'kg', 'weight unit kilograms', 'woocommerce' ),
```
There may not be an easy answer to this one, and really I'm just being mindful of the fact that some recent and otherwise valid i18n changes have inadvertently caused some measure of disruption: by switching from non-contextual to contextual strings, I think we'd effectively be throwing out the [previous translations](https://github.com/woocommerce/woocommerce/pull/36708/files#diff-7c6852728aa4a52b0f1420a2b6a886a902fb32230da64ca8415b9ab752cad32aL115). 

Is there a way to minimize the disruption that might cause for our merchants (not sure if there *is* a technical solution, but perhaps we could issue a CTA for community translators, via the dev blog)?

coreymckrill replied:
> I think we'd effectively be throwing out the [previous translations](https://github.com/woocommerce/woocommerce/pull/36708/files#diff-7c6852728aa4a52b0f1420a2b6a886a902fb32230da64ca8415b9ab752cad32aL115).

From [this](https://make.wordpress.org/polyglots/handbook/translating/glotpress-translate-wordpress-org/#strings-status-color-key), I think you're right. The strings would probably be changed to "fuzzy" status, so they'd have to be reapproved before they'd go back into the language pack.

Adding the context seems like the right thing to do here from a translator UX perspective, since something like "g" out of context doesn't mean much. But this might be better done in conjunction with other improvements around continuous deployment, where new and changed translations could get submitted to Polyglots right away instead of only when a new version is released.

I'll change it back to the simple `__()` function for now.

coreymckrill replied:
Changed back in 9691159

Note that this necessitates a slight change in the code snippet used for testing, updated above.

barryhughes reviewed (commented):
Nice, I like this. Left a couple of initial questions relating to back-compat.

coreymckrill committed:
Switch back to context-less gettext for unit translations

While the extra context would be beneficial for translators, these
particular strings have already been translated for many lanugages
and adding the context would invalidate the translations, which could
cause a visual regression for merchants if the new strings don't get
reapproved before the next version is released.

coreymckrill committed:
Return the input string if it's not a recognized unit of measure

Rather than returning an empty string as the label for an unknown
unit of measure, this allows a merchant to use a custom unit of measure
for something and still have it appear in the UI.

coreymckrill committed:
Clean up linting errors

barryhughes commented on plugins/woocommerce/tests/php/src/Utilities/I18nUtilTest.php:
```diff
@@ -8,43 +8,97 @@
+	 * @return string
+	 */
+	public static function filter_gettext( $translated ): string {
+		return sprintf( '%s (translated)', $translated );
```
Nice approach.

barryhughes commented:
Unrelated to your change, I notice the importer consistently uses `lbs` and `in` (even if the current units are `kg` and `mm`):

<img width="147" alt="import-placeholders" src="https://user-images.githubusercontent.com/3594411/219224380-c1011be4-7f58-45fb-b072-4aa0db832e89.png">

Will circle back and check if we have an existing issue for this, else will log a fresh issue.

barryhughes commented:
@coreymckrill this one is also not really in scope for your PR—but interested in your opinion (and I can log separately if we like):

<img width="662" alt="unit-selector" src="https://user-images.githubusercontent.com/3594411/219225240-55850e10-6dc9-47a5-98b5-9f7f66c0950f.png">

With the test snippet in place, we still see the untranslated units within **WooCommerce ▸ Products ▸ Settings ▸ Measurements** ... but it seems like perhaps that isn't ideal. What do you think—separate bug? Or not really a concern (certainly it has been this way for some time).

barryhughes commented on plugins/woocommerce/src/Utilities/I18nUtil.php:
```diff
@@ -0,0 +1,66 @@
+	 * Get the translated label for a weight unit of measure.
+	 *
+	 * This will return the original input string if it isn't found in the units array. This way a custom unit of
+	 * measure can be used even if its' not getting translated.
```
Final minor nit, sorry!

```suggestion
	 * measure can be used even if it's not getting translated.
```

barryhughes commented on plugins/woocommerce/src/Utilities/I18nUtil.php:
```diff
@@ -0,0 +1,66 @@
+	 * Get the translated label for a dimensions unit of measure.
+	 *
+	 * This will return the original input string if it isn't found in the units array. This way a custom unit of
+	 * measure can be used even if its' not getting translated.
```
Very minor/repeat of earlier note:

```suggestion
	 * measure can be used even if it's not getting translated.
```

barryhughes reviewed (dismissed):
LGTM!

Couple of minor but non-essential tweaks we could do (typos), otherwise this is good to go. I also left a couple of questions about something we may wish to deal with via a separate issue, but it's not a blocker for this work.

coreymckrill committed:
Update plugins/woocommerce/src/Utilities/I18nUtil.php

Co-authored-by: Barry Hughes <3594411+barryhughes@users.noreply.github.com>

coreymckrill committed:
Update plugins/woocommerce/src/Utilities/I18nUtil.php

Co-authored-by: Barry Hughes <3594411+barryhughes@users.noreply.github.com>

coreymckrill commented:
@barryhughes 

> With the test snippet in place, we still see the untranslated units within WooCommerce ▸ Products ▸ Settings ▸ Measurements ... but it seems like perhaps that isn't ideal.

Are you still seeing this if you update the testing snippet to match what's in the PR description? (I had to change it a little after I updated the translations to use `__()` instead of `_x()`)

I'm seeing the emojis when I look at those settings in my local dev...

barryhughes commented:
@coreymckrill yep, I switched from `'gettext_with_context'` to `'gettext'` (that's the change you mean, right?) and I did successfully see the emoji in the product editor's shipping tab, as well as the frontend single product page ... but not in the importer.

However, given my local settings (`kg` | `mm`) something is up that probably pre-dates your work, since I am still seeing Imperial units (`lbs` | `in`) in that screen.

coreymckrill commented:
> but not in the importer.

@barryhughes 🤔 Where in the importer are you seeing this? When I do an export of my products and then import that CSV file, here's what I see on the field mapping screen:

![import](https://user-images.githubusercontent.com/916023/219476349-736050f9-d9c9-4e37-8c6a-de44e4356b40.jpg)

barryhughes commented:
Err, let's just ignore me 🤦🏼 ... I was actually looking at the column headers (ie, the left hand side of your screenshot) which of course comes from the CSV itself.
